File: typescript/src/common/store.ts
import { randomBytes } from "crypto";

export type Merchant = {
  id: string;
  name: string;
  apiKey: string;
};

export type PaymentIntent = {
  id: string;
  merchantId: string;
  amount: number;
  currency: string;
  customerId?: string;
  paymentMethodToken: string;
  captureMethod: string;
  metadata: Record<string, string>;
  returnUrl?: string;
  status: string;
  clientSecret: string;
  chargeId?: string;
  processorReference?: string;
  createdAt: string;
  updatedAt: string;
};

export type Charge = {
  id: string;
  paymentIntentId: string;
  amount: number;
  currency: string;
  status: string;
  createdAt: string;
};

export type Refund = {
  id: string;
  chargeId: string;
  amount: number;
  status: string;
  createdAt: string;
};

export type WebhookEndpoint = {
  id: string;
  merchantId: string;
  url: string;
  enabled: boolean;
  createdAt: string;
};

export type Event = {
  id: string;
  merchantId: string;
  type: string;
  createdAt: string;
  data: Record<string, unknown>;
  status: string;
};

export type FraudAssessment = {
  id: string;
  paymentIntentId: string;
  riskScore: number;
  decision: string;
  signals: Record<string, unknown>;
  createdAt: string;
};

export type IdempotencyRecord = {
  requestHash: string;
  responseBody: string;
  statusCode: number;
  createdAt: string;
};

export type Task = { type: "process_payment" | "deliver_webhooks"; intentId?: string; eventId?: string };

const generateId = (prefix: string) => `${prefix}_${randomBytes(12).toString("hex")}`;
const utcNow = () => new Date().toISOString();

export class InMemoryStore {
  merchants = new Map<string, Merchant>();
  paymentIntents = new Map<string, PaymentIntent>();
  charges = new Map<string, Charge>();
  refunds = new Map<string, Refund>();
  idempotency = new Map<string, IdempotencyRecord>();
  webhookEndpoints = new Map<string, WebhookEndpoint>();
  events = new Map<string, Event>();
  fraudAssessments = new Map<string, FraudAssessment>();
  tasks: Task[] = [];

  addMerchant(name: string, apiKey: string): Merchant {
    const merchant: Merchant = { id: generateId("mch"), name, apiKey };
    this.merchants.set(merchant.id, merchant);
    return merchant;
  }

  getMerchantByApiKey(apiKey: string): Merchant | undefined {
    for (const merchant of this.merchants.values()) {
      if (merchant.apiKey === apiKey) return merchant;
    }
    return undefined;
  }

  getIdempotency(merchantId: string, key: string): IdempotencyRecord | undefined {
    return this.idempotency.get(`${merchantId}:${key}`);
  }

  setIdempotency(
    merchantId: string,
    key: string,
    requestHash: string,
    responseBody: string,
    statusCode: number
  ) {
    this.idempotency.set(`${merchantId}:${key}`, {
      requestHash,
      responseBody,
      statusCode,
      createdAt: utcNow()
    });
  }

  createPaymentIntent(merchantId: string, payload: Omit<PaymentIntent, "id" | "merchantId" | "status" | "clientSecret" | "createdAt" | "updatedAt">): PaymentIntent {
    const id = generateId("pi");
    const intent: PaymentIntent = {
      ...payload,
      id,
      merchantId,
      status: "requires_confirmation",
      clientSecret: `${id}_secret_${randomBytes(6).toString("hex")}`,
      createdAt: utcNow(),
      updatedAt: utcNow()
    };
    this.paymentIntents.set(id, intent);
    return intent;
  }

  getPaymentIntent(id: string): PaymentIntent | undefined {
    return this.paymentIntents.get(id);
  }

  updatePaymentIntent(id: string, updates: Partial<PaymentIntent>): PaymentIntent | undefined {
    const intent = this.paymentIntents.get(id);
    if (!intent) return undefined;
    const updated = { ...intent, ...updates, updatedAt: utcNow() };
    this.paymentIntents.set(id, updated);
    return updated;
  }

  createCharge(intentId: string, amount: number, currency: string, status: string): Charge {
    const charge: Charge = {
      id: generateId("ch"),
      paymentIntentId: intentId,
      amount,
      currency,
      status,
      createdAt: utcNow()
    };
    this.charges.set(charge.id, charge);
    return charge;
  }

  getCharge(id: string): Charge | undefined {
    return this.charges.get(id);
  }

  createRefund(chargeId: string, amount: number, status: string): Refund {
    const refund: Refund = {
      id: generateId("rf"),
      chargeId,
      amount,
      status,
      createdAt: utcNow()
    };
    this.refunds.set(refund.id, refund);
    return refund;
  }

  addWebhookEndpoint(merchantId: string, url: string, enabled: boolean): WebhookEndpoint {
    const endpoint: WebhookEndpoint = {
      id: generateId("we"),
      merchantId,
      url,
      enabled,
      createdAt: utcNow()
    };
    this.webhookEndpoints.set(endpoint.id, endpoint);
    return endpoint;
  }

  listWebhookEndpoints(merchantId: string): WebhookEndpoint[] {
    return [...this.webhookEndpoints.values()].filter(
      (endpoint) => endpoint.merchantId === merchantId && endpoint.enabled
    );
  }

  createEvent(merchantId: string, type: string, data: Record<string, unknown>): Event {
    const event: Event = {
      id: generateId("evt"),
      merchantId,
      type,
      createdAt: utcNow(),
      data,
      status: "pending"
    };
    this.events.set(event.id, event);
    this.tasks.push({ type: "deliver_webhooks", eventId: event.id });
    return event;
  }

  updateEventStatus(eventId: string, status: string) {
    const event = this.events.get(eventId);
    if (!event) return;
    this.events.set(eventId, { ...event, status });
  }

  listEvents(merchantId: string): Event[] {
    return [...this.events.values()].filter((event) => event.merchantId === merchantId);
  }

  createFraudAssessment(
    intentId: string,
    riskScore: number,
    decision: string,
    signals: Record<string, unknown>
  ) {
    const assessment: FraudAssessment = {
      id: generateId("fra"),
      paymentIntentId: intentId,
      riskScore,
      decision,
      signals,
      createdAt: utcNow()
    };
    this.fraudAssessments.set(assessment.id, assessment);
    return assessment;
  }

  enqueueTask(task: Task) {
    this.tasks.push(task);
  }

  dequeueTask(): Task | undefined {
    return this.tasks.shift();
  }
}

export const store = new InMemoryStore();

File: go/internal/store/store.go
package store

import (
	"crypto/rand"
	"encoding/hex"
	"sync"
	"time"

	"payment-processing/internal/models"
)

type InMemoryStore struct {
	mu               sync.Mutex
	Merchants        map[string]models.Merchant
	PaymentIntents   map[string]models.PaymentIntent
	Charges          map[string]models.Charge
	Refunds          map[string]models.Refund
	Idempotency      map[string]models.IdempotencyRecord
	WebhookEndpoints map[string]models.WebhookEndpoint
	Events           map[string]models.Event
	FraudAssessments map[string]models.FraudAssessment
	Tasks            []models.Task
}

func NewInMemoryStore() *InMemoryStore {
	return &InMemoryStore{
		Merchants:        map[string]models.Merchant{},
		PaymentIntents:   map[string]models.PaymentIntent{},
		Charges:          map[string]models.Charge{},
		Refunds:          map[string]models.Refund{},
		Idempotency:      map[string]models.IdempotencyRecord{},
		WebhookEndpoints: map[string]models.WebhookEndpoint{},
		Events:           map[string]models.Event{},
		FraudAssessments: map[string]models.FraudAssessment{},
		Tasks:            []models.Task{},
	}
}

func GenerateID(prefix string) string {
	return prefix + "_" + randomHex(12)
}

func randomHex(bytesLen int) string {
	b := make([]byte, bytesLen)
	_, _ = rand.Read(b)
	return hex.EncodeToString(b)
}

func utcNow() string {
	return time.Now().UTC().Format(time.RFC3339Nano)
}

func (s *InMemoryStore) AddMerchant(name, apiKey string) models.Merchant {
	s.mu.Lock()
	defer s.mu.Unlock()
	merchant := models.Merchant{
		ID:     GenerateID("mch"),
		Name:   name,
		APIKey: apiKey,
	}
	s.Merchants[merchant.ID] = merchant
	return merchant
}

func (s *InMemoryStore) GetMerchantByAPIKey(apiKey string) (models.Merchant, bool) {
	s.mu.Lock()
	defer s.mu.Unlock()
	for _, merchant := range s.Merchants {
		if merchant.APIKey == apiKey {
			return merchant, true
		}
	}
	return models.Merchant{}, false
}

func (s *InMemoryStore) GetIdempotency(merchantID, key string) (models.IdempotencyRecord, bool) {
	record, ok := s.Idempotency[merchantID+":"+key]
	return record, ok
}

func (s *InMemoryStore) SetIdempotency(merchantID, key, requestHash string, response []byte, statusCode int) {
	s.Idempotency[merchantID+":"+key] = models.IdempotencyRecord{
		RequestHash: requestHash,
		Response:    response,
		StatusCode:  statusCode,
		CreatedAt:   utcNow(),
	}
}

func (s *InMemoryStore) CreatePaymentIntent(merchantID string, payload models.PaymentIntent) models.PaymentIntent {
	s.mu.Lock()
	defer s.mu.Unlock()
	intent := payload
	intent.ID = GenerateID("pi")
	intent.MerchantID = merchantID
	intent.Status = "requires_confirmation"
	intent.ClientSecret = intent.ID + "_secret_" + randomHex(8)
	intent.CreatedAt = utcNow()
	intent.UpdatedAt = utcNow()
	s.PaymentIntents[intent.ID] = intent
	return intent
}

func (s *InMemoryStore) GetPaymentIntent(id string) (models.PaymentIntent, bool) {
	intent, ok := s.PaymentIntents[id]
	return intent, ok
}

func (s *InMemoryStore) UpdatePaymentIntent(id string, updates map[string]interface{}) (models.PaymentIntent, bool) {
	s.mu.Lock()
	defer s.mu.Unlock()
	intent, ok := s.PaymentIntents[id]
	if !ok {
		return models.PaymentIntent{}, false
	}
	if status, ok := updates["status"].(string); ok {
		intent.Status = status
	}
	if chargeID, ok := updates["charge_id"].(string); ok {
		intent.ChargeID = chargeID
	}
	if processorRef, ok := updates["processor_reference"].(string); ok {
		intent.ProcessorReference = processorRef
	}
	intent.UpdatedAt = utcNow()
	s.PaymentIntents[id] = intent
	return intent, true
}

func (s *InMemoryStore) CreateCharge(intentID string, amount int, currency, status string) models.Charge {
	s.mu.Lock()
	defer s.mu.Unlock()
	charge := models.Charge{
		ID:              GenerateID("ch"),
		PaymentIntentID: intentID,
		Amount:          amount,
		Currency:        currency,
		Status:          status,
		CreatedAt:       utcNow(),
	}
	s.Charges[charge.ID] = charge
	return charge
}

func (s *InMemoryStore) GetCharge(id string) (models.Charge, bool) {
	charge, ok := s.Charges[id]
	return charge, ok
}

func (s *InMemoryStore) CreateRefund(chargeID string, amount int, status string) models.Refund {
	s.mu.Lock()
	defer s.mu.Unlock()
	refund := models.Refund{
		ID:        GenerateID("rf"),
		ChargeID:  chargeID,
		Amount:    amount,
		Status:    status,
		CreatedAt: utcNow(),
	}
	s.Refunds[refund.ID] = refund
	return refund
}

func (s *InMemoryStore) AddWebhookEndpoint(merchantID, url string, enabled bool) models.WebhookEndpoint {
	s.mu.Lock()
	defer s.mu.Unlock()
	endpoint := models.WebhookEndpoint{
		ID:         GenerateID("we"),
		MerchantID: merchantID,
		URL:        url,
		Enabled:    enabled,
		CreatedAt:  utcNow(),
	}
	s.WebhookEndpoints[endpoint.ID] = endpoint
	return endpoint
}

func (s *InMemoryStore) ListWebhookEndpoints(merchantID string) []models.WebhookEndpoint {
	endpoints := []models.WebhookEndpoint{}
	for _, endpoint := range s.WebhookEndpoints {
		if endpoint.MerchantID == merchantID && endpoint.Enabled {
			endpoints = append(endpoints, endpoint)
		}
	}
	return endpoints
}

func (s *InMemoryStore) CreateEvent(merchantID, eventType string, data map[string]interface{}) models.Event {
	s.mu.Lock()
	defer s.mu.Unlock()
	event := models.Event{
		ID:         GenerateID("evt"),
		MerchantID: merchantID,
		Type:       eventType,
		CreatedAt:  utcNow(),
		Data:       data,
		Status:     "pending",
	}
	s.Events[event.ID] = event
	s.Tasks = append(s.Tasks, models.Task{Type: "deliver_webhooks", EventID: event.ID})
	return event
}

func (s *InMemoryStore) UpdateEventStatus(eventID, status string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	event := s.Events[eventID]
	event.Status = status
	s.Events[eventID] = event
}

func (s *InMemoryStore) ListEvents(merchantID string) []models.Event {
	events := []models.Event{}
	for _, event := range s.Events {
		if event.MerchantID == merchantID {
			events = append(events, event)
		}
	}
	return events
}

func (s *InMemoryStore) CreateFraudAssessment(intentID string, score int, decision string, signals map[string]interface{}) models.FraudAssessment {
	s.mu.Lock()
	defer s.mu.Unlock()
	assessment := models.FraudAssessment{
		ID:             GenerateID("fra"),
		PaymentIntentID: intentID,
		RiskScore:      score,
		Decision:       decision,
		Signals:        signals,
		CreatedAt:      utcNow(),
	}
	s.FraudAssessments[assessment.ID] = assessment
	return assessment
}

func (s *InMemoryStore) EnqueueTask(task models.Task) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.Tasks = append(s.Tasks, task)
}

func (s *InMemoryStore) DequeueTask() (models.Task, bool) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if len(s.Tasks) == 0 {
		return models.Task{}, false
	}
	task := s.Tasks[0]
	s.Tasks = s.Tasks[1:]
	return task, true
}

var Store = NewInMemoryStore()

